src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/WebhookGoCardlessService.java

```diff
@@ -63,22 +72,41 @@ public void handleEvents(List<GoCardlessEvent> events) {
 
     private void updateStateForMandateEvents(List<GoCardlessEvent> eventsThatAffectMandates) {
         eventsThatAffectMandates.stream()
-                .map(WebhookGoCardlessService::toGoCardlessMandateIdAndOrganisationId)
+                .map(this::toMandateBusinessKey)
                 .flatMap(Optional::stream)
                 .distinct()
+                .map(this::getMandate)
+                .flatMap(Optional::stream)
                 .forEach(goCardlessMandateStateUpdater::updateState);
     }
 
-    private static Optional<GoCardlessMandateIdAndOrganisationId> toGoCardlessMandateIdAndOrganisationId(GoCardlessEvent goCardlessEvent) {
+    private Optional<Pair<GoCardlessMandateId, GoCardlessOrganisationId>> toMandateBusinessKey(GoCardlessEvent goCardlessEvent) {
         var goCardlessMandateIdAndOrganisationId = goCardlessEvent.getLinksMandate()
-                .map(mandateId -> new GoCardlessMandateIdAndOrganisationId(mandateId, goCardlessEvent.getLinksOrganisation()))
+                .map(mandateId -> Pair.of(mandateId, goCardlessEvent.getLinksOrganisation()))
                 .orElseGet(() -> {
                     LOGGER.error("GoCardless event " + goCardlessEvent.getGoCardlessEventId() + " has resource_type mandate but no links.mandate");
                     return null;
                 });
         return Optional.ofNullable(goCardlessMandateIdAndOrganisationId);
     }
 
+    private Optional<Mandate> getMandate(Pair<GoCardlessMandateId, GoCardlessOrganisationId> goCardlessMandateBusinessKey) {
+        Mandate mandate = null;
+
+        try {
+            mandate = mandateQueryService.findByProviderMandateIdAndOrganisationId(
+                    GOCARDLESS,
+                    goCardlessMandateBusinessKey.getLeft(),
+                    goCardlessMandateBusinessKey.getRight());
+        } catch (MandateNotFoundException e) {
+            LOGGER.error(String.format("Could not update status of GoCardless mandate %s for organisation %s because the mandate was not found",
+                    goCardlessMandateBusinessKey.getLeft(),
+                    goCardlessMandateBusinessKey.getRight()));
```

Comment:
If we’re going to use a `Pair`, can we at least extract and name the fields first?

```java
var goCardlessMandateId = goCardlessMandateBusinessKey.getLeft();
var goCardlessOrganisationId = goCardlessMandateBusinessKey.getRight();
try {
    mandate = mandateQueryService.findByProviderMandateIdAndOrganisationId(GOCARDLESS, goCardlessMandateId, goCardlessOrganisationId);
} catch (MandateNotFoundException e) {
    LOGGER.error("…", goCardlessMandateId, goCardlessOrganisationId);
}
```